[Calling]: Feature : maximize/minimize video/screen share on double click

mejdoo opened the pull request:
## What's new in this PR?

Once this PR is merged, it will be possible during a call to maximise a video or screen share of any user on  double click.

**No need anymore to turn off all the videos if someone is sharing the screen.**

## Note:

The feature flag `callingVvmMaximizeMinimizeVideo`  is now set to **true** for testing purposes, till we decide either this feature will go with next release train or not. 

https://wearezeta.atlassian.net/browse/SQCALL-9
#### APK
[Download build #2964](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2964/artifact/build/artifact/wire-dev-PR3097-2964.apk)
[Download build #2965](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2965/artifact/build/artifact/wire-dev-PR3097-2965.apk)
[Download build #2968](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2968/artifact/build/artifact/wire-dev-PR3097-2968.apk)
[Download build #2970](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2970/artifact/build/artifact/wire-dev-PR3097-2970.apk)
[Download build #2978](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2978/artifact/build/artifact/wire-dev-PR3097-2978.apk)
[Download build #2979](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2979/artifact/build/artifact/wire-dev-PR3097-2979.apk)
[Download build #2981](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2981/artifact/build/artifact/wire-dev-PR3097-2981.apk)
[Download build #2982](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2982/artifact/build/artifact/wire-dev-PR3097-2982.apk)
[Download build #2984](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2984/artifact/build/artifact/wire-dev-PR3097-2984.apk)

mejdoo commented on app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala:
```diff
@@ -250,62 +174,36 @@ class CallingFragment extends FragmentHelper {
-
-  override def onDestroyView(): Unit = {
-    super.onDestroyView()
+  def clearVideoGrid(): Unit = {
```
The VideoGrid should be cleared and reloaded otherwise AVS native lib will be crashing. There is a limitation right now, we can't use `com.waz.avs.VideoRenderer` with same parameters in different views, at the same time.

BradleyWire commented on app/src/main/scala/com/waz/zclient/calling/FullScreenVideoFragment.scala:
```diff
@@ -0,0 +1,94 @@
+  val PARTICIPANT_BUNDLE_KEY = "participant"
+
+  def newInstance(participant: Participant): Fragment = {
+    val fragment = new FullScreenVideoFragment
```
You can use the `returning(new FullScreenVideoFragment) { fragment =>  }` here to clean it up a little

mejdoo replied:
ok done

BradleyWire commented on app/src/main/scala/com/waz/zclient/calling/views/UserVideoView.scala:
```diff
@@ -0,0 +1,125 @@
+  Signal.zip(controller.isGroupCall, controller.controlsVisible,
+    controller.otherParticipants.map(_.size)
+  ).map {
+    case (true, false, 0 | 1 | 2) => View.GONE
```
formatting here the => need to line up

makingthematrix replied:
And also either please put all three original signals in one line, or each in a separate line.

BradleyWire commented on app/src/main/scala/com/waz/zclient/calling/FullScreenVideoFragment.scala:
```diff
@@ -0,0 +1,94 @@
+}
+
+object FullScreenVideoFragment {
+  val Tag = classOf[FullScreenVideoFragment].getName
```
Do we want this like this? In case of obfuscation this may not display the correct tag

mejdoo replied:
I just followed what it was done in other fragments.

makingthematrix replied:
There was never a problem with this, to be honest. I'm fine with this.

gizemb replied:
we don't obfuscate the code

BradleyWire commented on app/src/main/scala/com/waz/zclient/calling/FullScreenVideoFragment.scala:
```diff
@@ -0,0 +1,94 @@
+      container.addView(userVideoView)
+
+      controller.allVideoReceiveStates
+        .map(_.getOrElse(participant, VideoState.Unknown)).onUi {
```
Is this formatted properly? the indentation looks off for the `case` values

BradleyWire commented on app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala:
```diff
@@ -17,160 +17,76 @@
-  protected val pausedTextVisible = stateMessageText.map(_.exists(_.nonEmpty))
-  pausedTextVisible.onUi(pausedText.setVisible)
+  private val maxVideoPreviews = 12
+  private lazy val controller = inject[CallController]
```
If we're grouping these values together, we normally align the `=`

mejdoo replied:
Is there anyway to define auto formatting rule for Scala in Android studio? Those classed are mostly auto formatted... @makingthematrix any idea?

makingthematrix replied:
This alignment is not automatic. It's just for readability and it's up to the person writing the code. If I see that something can be easier to write with some small changes in alignment then I do it, but it's not a requirement.

BradleyWire commented on app/src/main/scala/com/waz/zclient/calling/views/UserVideoView.scala:
```diff
@@ -0,0 +1,125 @@
+  protected def registerHandler(view: View) = {
+    controller.allVideoReceiveStates.map(_.getOrElse(participant, VideoState.Unknown)).onUi {
+      case VideoState.Paused | VideoState.Stopped => view.fadeOut()
+      case _                 => view.fadeIn()
```
formatting here too

gizemb commented on app/src/main/scala/com/waz/zclient/calling/FullScreenVideoFragment.scala:
```diff
@@ -0,0 +1,94 @@
+
+      val selfParticipant = controller.callingZms.map(zms => Participant(zms.selfUserId, zms.clientId)).currentValue.get
+
+      var userVideoView: UserVideoView = null
+      if (participant == selfParticipant) userVideoView = new SelfVideoView(getContext, participant)
+      else userVideoView = new OtherVideoView(getContext, participant)
```
this doesn't need to be a `var` I guess?
`val userVideoView = if (...) SelfVideoView() else OtherVideoView()`

makingthematrix replied:
I think in general `userVideoView` should be a lazy val defined on the class level and in its initialization block we can check the bundle for info what should it be and should it subscribe to any signal.

mejdoo replied:
@makingthematrix  I didn't get how you wanna refactor this. 

I think @gizemb proposal looks fine.

gizemb commented on app/src/main/scala/com/waz/zclient/calling/FullScreenVideoFragment.scala:
```diff
@@ -0,0 +1,94 @@
+      }
+    }
+
+  def minimizeVideo(container: FrameLayout, userVideoView: UserVideoView): Unit = {
+    container.removeView(userVideoView)
+    getFragmentManager.popBackStack()
+    getParentFragment.asInstanceOf[CallingFragment].initVideoGrid()
+  }
```
there's a really tight coupling between this fragment and CallingFragment. don't we have any other way to communicate?

makingthematrix replied:
An event stream defined in a controller connecting the two fragments could help:
```
val initVideo = EventStream[Unit]() // in CallingController, for example
initVideo.foreach(callingFragment.initVideoGrid)
...
callingController.initVideo ! (()) // here in FullScreenVideoFragment
```

mejdoo replied:
ok done!

makingthematrix commented on app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala:
```diff
@@ -17,160 +17,76 @@
 
-  protected val pausedTextVisible = stateMessageText.map(_.exists(_.nonEmpty))
-  pausedTextVisible.onUi(pausedText.setVisible)
+  private val maxVideoPreviews = 12
```
Constants should be defined in the companion object, and the name should start with the capital letter.

mejdoo replied:
so not all the name should be capitalised like Java or Kotlin style?

makingthematrix replied:
@mejdoo : You can see `CONSTANT_LIKE_THIS` too, but at least for me this style means this is aconstant that should be resolved at compile time, and `ConstantLikeThis` is just a magic number, a default string, etc. I agree it's not consistent.

mejdoo replied:
ok anyway, I changed this to `MaxVideoPreviews` in the companion object.

makingthematrix commented on app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala:
```diff
@@ -17,160 +17,76 @@
-    _.setBackgroundColor(getColor(R.color.black_16))
+  override def onCreate(savedInstanceState: Bundle): Unit = {
+    super.onCreate(savedInstanceState)
+    controller.theme.map(themeController.getTheme).onUi(theme => videoGrid.foreach(_.setBackgroundColor(getStyledColor(R.attr.wireBackgroundColor, theme))))
```
Instead of subscribing to the signal here, put it in the `videoGrid` definition:
```
private lazy val videoGrid = returning(view[GridLayout](R.id.video_grid)) { vh =>
  controller.theme.map(themeController.getTheme).onUi { theme =>
    vh.foreach { ... }
  }
}
```

Tak a look at other places this pattern is used.

makingthematrix commented on app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala:
```diff
@@ -179,18 +95,26 @@ class CallingFragment extends FragmentHelper {
+      } { v =>
+        viewMap = viewMap.updated(participant, v)
+        if (BuildConfig.CALLING_VVM_MAXIMIZE_MINIMIZE_VIDEO) {
+          v.onDoubleClick.onUi { _ =>
```
Subscribing to a signal or an event stream inside a method may sometimes lead to unclear behaviour. It's more readable if `v.onDoubleClick.onUi { _ => ... }` is defined on the main level and then, inside the `onUi{ _ => ... }` the app checks what should happen because of the click. But I'm not sure yet how to do it in this particular case. I can look more into it on Monday.

makingthematrix commented on app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala:
```diff
@@ -179,18 +95,26 @@ class CallingFragment extends FragmentHelper {
+      val isVideoBeingSent = !vrs.get(selfParticipant).contains(VideoState.Stopped)
 
-      vh.foreach { v =>
+      videoGrid.foreach { v =>
```
This can also be moved to the definition of `videoGrid`.

mejdoo replied:
I moved this block outside of the videoGrid lazy loading declaration because I need to clear and initialize the videoGrid multiple times when I maximize/minimize any idea. Otherwise, AVS lib will crash.

makingthematrix commented on app/src/main/scala/com/waz/zclient/calling/FullScreenVideoFragment.scala:
```diff
@@ -0,0 +1,94 @@
+
+      container.addView(userVideoView)
+
+      controller.allVideoReceiveStates
```
This subscription to the signal should also be on the class level. Generally, if for some reason it's not to difficult, the logic should be inverted: First subscribe, then inside the subscription decide what to do. Or, even better, you can use `.filter`, `.map`, `.flatMap`, and `.collect` to build the logic into the signal or event stream.

mejdoo replied:
Not sure how to do it in this case, as I need to call `minimizeVideo(container, userVideoView)` with the parameter  `userVideoView` created inside `showMaximizedVideo(..)`.

makingthematrix commented on app/src/main/scala/com/waz/zclient/calling/views/UserVideoView.scala:
```diff
@@ -0,0 +1,125 @@
+    }
+    view match {
+      case vr: VideoRenderer =>
+        controller.allVideoReceiveStates.map(_.getOrElse(participant, VideoState.Unknown)).onUi {
```
Same issue as before - it would be better to have a subscription to a signal on the class level, and logic what to do when the signal changes implmeneted inside the subscription, or through map, filter, etc.

mejdoo replied:
This is an old implementation, I just moved the classes to separate files. Could you provide a code snippet for the refactoring.  In this case,`registerHandler()` is called from the classes `SelfVideoView` and `OtherVideoView` which extends from `UserVideoView`.

makingthematrix commented on zmessaging/src/main/scala/com/waz/service/call/CallInfo.scala:
```diff
@@ -143,7 +143,7 @@ case class CallInfo(convId:             ConvId,
 object CallInfo {
 
-  case class Participant(userId: UserId, clientId: ClientId, muted: Boolean = false)
+  case class Participant(userId: UserId, clientId: ClientId, muted: Boolean = false) extends Serializable
```
Do you really need to extend Serializable? (btw, I'm not sure, but a case class should do it already, I think...)

mejdoo replied:
right! not needed :)

mejdoo commented on app/build.gradle:
```diff
@@ -123,9 +123,10 @@ android {
+        //Feature Flags
         buildConfigField 'boolean', 'KOTLIN_SETTINGS', "$rootProject.ext.kotlinSettings"
         buildConfigField 'boolean', 'KOTLIN_CORE', "$rootProject.ext.kotlinCore"
+        buildConfigField 'boolean', 'CALLING_VVM_MAXIMIZE_MINIMIZE_VIDEO', "$rootProject.ext.callingVvmMaximizeMinimizeVideo"
```
I wanted to specify the squad and the parent feature name. Anyway, `MAXIMIZE_MINIMIZE_VIDEO` would be enough.

makingthematrix commented:
I pushed a commit with changes. My main issue was with this very big and complicated method, `initVideoGrid` which created a signal and subscribed to it - that could potentially lead to memory leaks. I tried to split the signal into a few smaller ones and connect them only later, and also move the subscription to the videoGrid `ViewHolder`. This way, the subscription can be managed automatically and for sure there will be only one at the time.

I still don't like that we subscribe to double clicks inside the method, but I think there's no risk in that.

mejdoo commented:
> I pushed a commit with changes. My main issue was with this very big and complicated method, `initVideoGrid` which created a signal and subscribed to it - that could potentially lead to memory leaks. I tried to split the signal into a few smaller ones and connect them only later, and also move the subscription to the videoGrid `ViewHolder`. This way, the subscription can be managed automatically and for sure there will be only one at the time.
> 
> I still don't like that we subscribe to double clicks inside the method, but now I think about it, there shouldn't be any risk in that...
> 
> I also changed a few things in `CallController`. Many signals there are initialized at once even if they are never used because there's no video call. I added `lazy` to their declarations.

Ok thank you. Let me test again the feature before merging...

makingthematrix commented on app/src/main/scala/com/waz/zclient/calling/FullScreenVideoFragment.scala:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  def showMaximizedVideo(participant: Participant): Unit =
+    fullScreenVideoContainer.foreach { container =>
```
Similar to the issue in `CallingFragment`:
If you make subscriptions like this inside methods, you risk memory leaks and unpredicted behaviour. Every call to the method will create a new subscription and they all will be called next time the signal refreshes, so the same logic will be executed many times when it should be executed only once.

Instead, make subscriptions to signals on the class level (or, in this case, inside `returning(view[FrameLayout](R.id.full_screen_video_container)) { ... }`), and let the method call update some data which in turn will make the signal refresh. 

For example, as I understand, we can have only one participant with maximized video. You can then create a source signal:
```
val maximizedParticipant = Signal[Option[Participant]]()
```
And `selfParticipant` is already a signal in the controller: `controller.selfParticipant`.
So, inside that `returning` clause you can zip together all three:
```
Signal.zip(
  maximizedParticipant,
  controller.selfParticipant,
  controller.allVideoReceiveStates
).foreach { ... }
```
And there you can choose to minimize and maximize views.

makingthematrix commented on app/src/main/scala/com/waz/zclient/calling/FullScreenVideoFragment.scala:
```diff
@@ -0,0 +1,93 @@
+      val selfParticipant = controller.callingZms.map(zms => Participant(zms.selfUserId, zms.clientId)).currentValue.get
+
+      val userVideoView = if (participant == selfParticipant) new SelfVideoView(getContext, participant)
+      else new OtherVideoView(getContext, participant)
```
I'd suggest you create a var `Map[Participant, UserVideoView]` and create new views only if they're not already there.

Also, I'm not sure about comparisons of the whole `Participant` case class. Aren't we only interested in user ids of participants?

makingthematrix commented on app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala:
```diff
@@ -17,298 +17,229 @@
-      case Some(p)             => view.setText(p.displayName)
-      case _                   =>
+      controller.initVideo.foreach { _ =>
+        initGridLayout()
```
`initGridLayout()` creates a new subscription every time it's called. I don't think this is what you want here. Instead, only one subscription to the signal should be made, in line 54, and here you only refresh the grid once. That means, you take `userGridInfo.head` and call `refreshGridLayout` with its data. 
So, something like this:
```
videoGridInfo.head.foreach { case (selfParticipant, videoUsers, infos, participants, isVideoBeingSent) =>
  controller.isFullScreenEnabled.head.collect { false =>
    vh.foreach { grid =>
      refreshVideoGrid(grid, selfParticipant, videoUsers, infos, participants, isVideoBeingSent)
     }
  }
}
```
It looks similar to `initGridLayout`, but the calls to `.head` make all the difference.

makingthematrix commented on app/src/main/scala/com/waz/zclient/calling/CallingFragment.scala:
```diff
@@ -17,298 +17,229 @@
+    def initGridLayout() =  videoGridInfo.foreach {
+      case (selfParticipant, videoUsers, infos, participants, isVideoBeingSent) =>
+        vh.foreach { grid =>
+          val isFullScreenEnabled = controller.isFullScreenEnabled.currentValue.getOrElse(false)
```
Don't access the value of the signal like this. Here you can zip those two signals instead: 
```
Signal.zip(videoGrodInfo, controller.isFullScreenEnabled).foreach {
  case ((sp, vu, info, ps, is), false) => ...
  case _ => 
}
```

AndroidBob commented:
Build [5286](https://10.10.124.134/job/AR-build-pipeline/5286/) **failed**.

AndroidBob commented:
Build [5288](https://10.10.124.134/job/AR-build-pipeline/5288/) **failed**.

AndroidBob commented:
Build [5290](https://10.10.124.134/job/AR-build-pipeline/5290/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.8.0-9931112-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.8.0-9931112-staging-compat.apk)
